common/src/Renderer/EntityDecalRenderer.cpp

```diff
@@ -0,0 +1,438 @@
+                         != data.brushes.end();
+
+    // if this brush is tracked by this entity or intersects, we'll need to
+    // recalculate the geometry
+    if (intersects || tracked)
+    {
+      invalidateDecalData(data);
+    }
+  }
+}
+
+void EntityDecalRenderer::removeBrush(const Model::BrushNode* brushNode)
+{
+  // invalidate any entities that are tracking this brush
+  for (auto& [ent, data] : m_entities)
+  {
+    // skip entities that are going to be recomputed anyway
+    if (!data.validated)
+    {
+      continue;
+    }
+
+    // if this brush is tracked by this entity, remove it and recalculate
+    const auto tracked = std::find(data.brushes.begin(), data.brushes.end(), brushNode)
+                         != data.brushes.end();
+    if (tracked)
+    {
+      invalidateDecalData(data);
+    }
+  }
+}
+
+void EntityDecalRenderer::invalidateDecalData(EntityDecalData& data) const
+{
+  // do nothing if the brush data is already marked as invalidated
+  if (!data.validated)
+  {
+    return;
+  }
+
+  data.validated = false;
+
+  // if the texture doesn't exist, do nothing
+  // also do nothing if the VBO storage fields are null, but it shouldn't happen
+  if (!data.texture || data.vertexHolderKey == nullptr || data.faceIndicesKey == nullptr)
+  {
+    return;
+  }
+
+  // update the VBO
+  m_vertexArray->deleteVerticesWithKey(data.vertexHolderKey);
+
+  const auto faceIndexHolder = m_faces->at(data.texture);
+  faceIndexHolder->zeroElementsWithKey(data.faceIndicesKey);
+
+  if (!faceIndexHolder->hasValidIndices())
+  {
+    // there are no indices left to render for this texture
+    m_faces->erase(data.texture);
+  }
+
+  data.vertexHolderKey = nullptr;
+  data.faceIndicesKey = nullptr;
+}
+
+void EntityDecalRenderer::validateDecalData(
+  const Model::EntityNode* entityNode, EntityDecalData& data) const
+{
+  if (data.validated)
+  {
+    // already validated, no need to check again
+    return;
+  }
+
+  const auto spec = getDecalSpecification(entityNode);
+  if (spec == std::nullopt)
+  {
+    // we shouldn't get here
+    data.validated = true;
+    return;
+  }
+
+  const auto& document = kdl::mem_lock(m_document);
+  const auto& editorContext = document->editorContext();
+  const auto* world = document->world();
+
+  // collect all the brush nodes that touch the entity's bbox
+  const auto bounds = entityNode->physicalBounds();
+  const auto intersectors = world->nodeTree().find_intersectors(bounds);
+
+  // track them in the entity
+  data.brushes.clear();
+  for (const auto* node : intersectors)
+  {
+    if (const auto* br = dynamic_cast<const Model::BrushNode*>(node))
+    {
+      // ignore hidden brushes
+      if (editorContext.visible(br))
+      {
+        data.brushes.push_back(br);
+      }
+    }
+  }
+
+  // shrink the bounds just a tiny bit so we get exclusive checks instead of inclusive
+  const auto bbox = entityNode->physicalBounds().expand(-0.01);
```

Comment:
I would consider writing a separate function to perform this check as it is done in the engine. My experience is that these little differences do turn up in bug reports later on, but by then we'll have forgotten all about how this works and we'll have to invest time to debug it. In the very least, explain this in a comment - but my recommendation would still be to write a separate implementation of `intersect_bbox_polygon` here. It can use the other check and just additionally check the corners - shouldn't cost much.

Reply:
Regarding "almost zero" float constants, we have `vm::C::almost_zero()` and a few more like these for that, check `vecmath/constants.h`. It's even properly documented :-P

Reply:
Hm, looking at it now, creating an exclusive version of the `intersect_bbox_polygon` method would need quite a bit of extra code to be added as it would also need exclusive versions of `bbox.contains`, `polygon_contains_point`, and `detail::handle_polygon_edge_intersection`, all of which use inclusive tests. I don't think it's as easy as calling the inclusive version of the method and then just adding some extra checks afterwards. I'll use the other suggestion of explaining the details in a comment, but feel free to correct me if I'm wrong.